spec/index.html

```diff
@@ -559,27 +556,28 @@ <h3>Triples</h3>
       |p| is called the <dfn>predicate</dfn> of the triple, and
       |o| is called the <dfn>object</dfn> of the triple.</p>
 
-    <p>A <dfn data-local-lt="quoted">quoted triple</dfn> is an <a>RDF triple</a> that is
-      used as the <a>subject</a> or <a>object</a> of another triple.</p>
+    <p>A <dfn data-local-lt="quoted">quoted graph</dfn> is an <a>RDF graph</a> that is
+      used as the <a>subject</a> or <a>object</a> of triple.</p>
 
     <p class="note">By the given definitions,
       the <a>subject</a> of an <a>RDF triple</a> may be either an <a>IRI</a>,
-      a <a>blank node</a>, or a <a>quoted triple</a>;
+      a <a>blank node</a>, or a <a>quoted graph</a>;
       the <a>predicate</a> of an <a>RDF triple</a> may only be an <a>IRI</a>;
       the <a>object</a> of an <a>RDF triple</a> may be either an <a>IRI</a>,
-      a <a>blank node</a>, a <a>literal</a> or a <a>quoted triple</a>.</p>
+      a <a>blank node</a>, a <a>literal</a> or a <a>quoted graph</a>.</p>
 
-    <p class="note">The definition of <a>quoted triple</a> is recursive.
-      That is, a <a>quoted triple</a> can itself have a
-      <a>subject</a> or <a>object</a> component which is another <a>quoted triple</a>.
-      However, by this definition, cycles of <a>quoted triples</a> cannot be created.</p>
+    <p class="note">The definition of <a>quoted graph</a> is recursive.
+      That is, a <a>quoted graph</a> can itself have a
+      triples where the <a>subject</a> or <a>object</a> components
+      are <a>quoted graphs</a>.
+      However, by this definition, cycles of <a>quoted graphs</a> cannot be created.</p>
```

Comment:
I don't expect implementations will always be able to detect, and hence error on, such cycles, until they are revealed by a SPARQL or other query. Thus, better to make the statement a forbiddance, rather than an impossibility.
```suggestion
      However, by this definition, cycles of <a>quoted graphs</a> are not permitted.</p>
```

Reply:
The way the rules work means cycles are impossible. The rules start with RDF 1.1 s-p-o triples (no triple terms) and iteratively allow triples with triple terms, first 1-deep nesting, then 2-deep nesting, ...

Because there is no reference to a triple term before it is created, cycles can't be built. (c.f. "let" and "letrec" in Lisp).

To me, "not permitted" suggests there is a route, but it's banned, which is not the case.

Reply:
I agree with Andy.

Reply:
I agree with Andy. Cycles are just not possible in the abstract syntax.

Of course, in concrete serialization formats that use some form of locally-scoped identifiers for graph terms (such as `%{g1}` [as proposed by Andy on the mailing list](https://lists.w3.org/Archives/Public/public-rdf-star-wg/2023Oct/0017.html)), serializations may contain cycles that need to be detected and raised as an error. But that's a topic for these serialization formats, not for the discussion of the abstract syntax in RDF Concepts.

Reply:
If we use some form of graph identifier to denote the graph, then if used as an object, there is some sense that one graph can “contain”another by reference. This could lead to a graph referencing itself, or through some indirection. If the notion is that graphs are strictly terms, and graph identifiers are only used to express this in a concrete syntax, then cycles are not possible. I’m not clear that we’ve settled on this, but I don’t think there’s a use case for graphs to reference themselves.

Reply:
The syntax USING-FOR can't create cycles if we say that the representative token `%{g}` can't be used in the definition `FOR {..}`.

N-Quads is harder to get this prescriptive effect.

Reply:
@afs If I understand your proposal for the syntax correctly, the `FOR {..}` part may contain such tokens (assuming graph terms may be nested). For instance:
```
USING %{g1} FOR { :s :p :o }
USING %{g2} FOR { %{g1} :p :o }
```
Right?
In this case, there can be cycles in this serialization format:
```
USING %{g1} FOR { %{g2} :p :o }
USING %{g2} FOR { %{g1} :p :o }
```
Yet, these cycles can be prevented if such a token is not allowed to be used inside any `FOR {..}` part before it is declared (i.e., with this requirement, the mention of `%{g2}` in the first line of my example above would be disallowed).

Reply:
I was assuming we didn't allow use-before-definition - that stops cycles. It is a significant change to TriG/Turtle.

However, there is another kind of cycle.

For any proposal with occurrences that are RDF resources, whether blank graph style or graph term based, it's unavoidable to have occurrence-cycles. 

```
  ## Blank graphs.
  _:a { :s :p _:a }
```
or
```
  ## Declared graph terms
  USING %{g1} FOR { :s :p _:a }
  _:a rdf:occurenceOf %{g1}
```

Reply:
Graphs (as immutable triple sets) forming descriptions of their occurrences is fine (e.g. records or minds are occasionally quite self-descriptive). It's just referencing (opaquely, only IRIs or blankIDs), so there is no structural/syntactic cycle. Which is as you say is impossible; a set cannot contain *itself*, only references denoting it.

And I take it that these `%{g}` markers aren't references, more like preprocessor substitutions (i.e. like inline definitions / `#define`,  not addresses)? (I took them as only used to flatten out graph terms into a quad structure, and not necessarily to be used in TriG? In SPARQL I can see more utility of that though.)

Reply:
Yes, `%{g}` are syntax only. Syntactic substitutions. Turtle doesn't allow the same syntactic unit to be shared in the object position. There is no equivalent of same-subject syntax in `:s :p 1; :p 2`. Writing or reading even 10 triple graph terms repeatedly is laborious.

Do you have a concrete example of an occurrence semantically referencing itself?

Reply:
Presumably `%{g}` is an aspect only of N-Quads and not available for use in TriG (I'm not in love with that syntax, but it will do for now). We could consider that graph terms can be used only in N-Quads and TriG, although we could consider extending Turtle to support graph terms.

I don't see a way to allow it in N-Triples without a fully expanded representation (which is possible), and it's not worth adding anything to RDF/XML. But, I think adding support to N-Triples would mean that it's no longer a strict subset of N-Quads.

We do have a nomenclature problem with Graph vs. Dataset. We define a Dataset as a default graph plus zero or more named graphs. By this definition datasets aren't required for graph terms, but the need for N-Quads to support `%{g}` conflates the two.

Reply:
I have tried not to say one name or the other, for the WG to decide.
Other proposals have used TriG because they exploit the TriG Syntax and not give a name to the token-type relationship. They then give rules for blank node use (e.g. not in `GRAPH ?g` or the union graph). There will interactions of syntax and semantics via entailment.

[2023Oct/0017](https://lists.w3.org/Archives/Public/public-rdf-star-wg/2023Oct/0017.html) started as an attempt to separate syntax from semantics.

The `%{g}` syntax doesn't have to be used in T* syntaxes (Turtle, TriG). They have same-subject syntax but no same-object so same-object requires writing out again and reading carefully.

One case is several occurrences of the same graph. Using rdf:type to give the nature of the occurrence:
```
_:a1 rdf:occurenceOf { :s :p :o } .
_:a1 rdf:type ... 
_:a2 rdf:occurenceOf { :s :p :o } .
_:a2 rdf:type ... 
```

If `{ :s :p :o }` has more triples, even as low as 10s, it's visually burdensome to write or read full graph terms for 5 occurrences of the same graph term (same by equality of sets of triples).

```
USING %{g} FOR { ... 10s of triples ... }
_:a1 rdf:occurenceOf %{g} 
_:a1 rdf:type ... 
_:a2 rdf:occurenceOf %{g} 
_:a2 rdf:type ... 
```
It's a shorthand. Blank node graphs proposals have the same issue. Some of the examples carefully put the graph in the subject position.

Reply:
> Yes, `%{g}` are syntax only. Syntactic substitutions. Turtle doesn't allow the same syntactic unit to be shared in the object position. There is no equivalent of same-subject syntax in `:s :p 1; :p 2`. Writing or reading even 10 triple graph terms repeatedly is laborious.

Yes, I agree. This is also related to why I proposed 1) a [new shorthand for unasserted triples/graphs](https://gist.github.com/niklasl/4f52c32ef2d888c172c8584e36c24610#proposal-new-quoted-form), re-using the  annotation syntax, but preceding objects with a "quotation dash" (or an SQL comment if you will), and also 2) adjusting the annotation syntax to be able to [describe multiple occurrences](https://gist.github.com/niklasl/4f52c32ef2d888c172c8584e36c24610#proposal-rdf-star-annotation-occurrences) of the same pattern, along with using identifiers of the occurrence (i.e. named graph). I haven't pushed more on that as we've been in the fundaments; but once we're "out of the weeds" I think it is important to consider. I agree that we must cater for the ergonomics in the concrete syntaxes. (I've been leaning towards supporting these advanced concepts only in TriG/N-Quads; but that depends on other factors.)

I [toyed with that syntax](https://gist.github.com/niklasl/2d02902b81e215b1795981df31927e9b) on all our collected use cases recently. I know this mightn't fly, but I particularly found the data around line [40](https://gist.github.com/niklasl/2d02902b81e215b1795981df31927e9b#file-ucr-examples-blankgraphs-alt-sugar-ttl-L40) and [67](https://gist.github.com/niklasl/2d02902b81e215b1795981df31927e9b#file-ucr-examples-blankgraphs-alt-sugar-ttl-L67) interesting in this variant form. I'm using an adjusted TriG parser supporting this notation, and have done some small tests of turning that into blank graphs, and recently into "fake graph terms" using those "specially named graphs" I [mentioned recently on the list](https://lists.w3.org/Archives/Public/public-rdf-star-wg/2023Oct/0018.html).

(The control options for what goes in the [union graph](https://patterns.dataincubator.org/book/union-graph.html) may leave this wanting though, since there is no real standard and lots of [implementation variation](https://timeaturdean.com/cross-rdf-graph/). Alas I haven't used Fuseki much in the near past, but I recall it having useful [configuration options](https://jena.apache.org/documentation/fuseki2/fuseki-configuration.html) here. I had that in the back of my mind when I thought of [explicit dataset composition](https://gist.github.com/niklasl/c22994e664663b6730613ecc1321c418#explicit-dataset-composition).)

 
> Do you have a concrete example of an occurrence semantically referencing itself?

I've seen (DCAT) Datasets being "part of themselves". And if "referencing" also means used in any position of a triple in the graph it is an occurrence of, I'd say the majority of all web pages are actually examples, since they at least use some `<link rel=stylesheet ... >`, and if you use the page URL as a name for the graph: `<page> { <page> xht:stylesheet <style.css> }`, I think that qualifies?  (Also, specifically in our union catalogue, each thing (book, person, topic, etc.) is described by a record which also describes itself ([example](https://libris.kb.se/xv8b95hg5khnh36/data.ttl?embellished=false)), and each record URI is used as a name for its graph in our quad store ([endpoint](https://libris.kb.se/sparql)).)

Reply:
> The `%{g}` syntax doesn't have to be used in T* syntaxes (Turtle, TriG). They have same-subject syntax but no same-object so same-object requires writing out again and reading carefully.

Perhaps another thing to consider importing from N3 is the `is ... of` construct, which affords the `:o is :p of :s` structure, which would allow the same-object case without resorting to `%{g}`. JSON-LD has `@reverse` which can be a way of doing the same thing. But that's looking like a syntax-driven solution. If we introduce `%{g}` in one syntax, it may be hard to not include it in others.

Reply:
I don't see a cycle in [example](https://libris.kb.se/xv8b95hg5khnh36/data.ttl?embellished=false). The record xv8b95hg5khnh36 does not reference itself.

> if you use the page URL as a name for the graph

Isn't that an indirect relationship? 
It would help me to see the N-quads-ish realisation of that convenience syntax. I'm not clear what's the type and what's the occurrence.

Reply:
> Perhaps another thing to consider importing from N3 is the `is ... of` construct, which affords the `:o is :p of :s` structure, which would allow the same-object case without resorting to `%{g}`. JSON-LD has `@reverse` which can be a way of doing the same thing.

"is..of" is useful when the declaration style of properties is that way round (an ontology I work with currently does use the "uplink" type of properties and "is ... of" would be nice to have). "is..of" might work for hand written input; I'm not sure it covers all cases.

For output at scale, a writer would have to inspect the whole file before starting to write as is necessary for use of `[ ]` syntax to nest blank nodes.

A simple RDF 1.1 Turtle writer is to output subject, then output all property,object pairs.
A very simple RDF 1.1 Turtle writer is output each subject/predicate/object in abbreviated form.

Simple writers output `_:label` rather than `[ ]`.

> If we introduce `%{g}` in one syntax, it may be hard to not include it in others.

Yes, especially if we go the Turtle route because N-Triples is a subset of Turtle.

As a syntax form in Turtle, using `%{g}` is mostly optional.

Reply:
Some syntax we might consider is `'{' ( '=' PN_CHARS+ )? triplesBlock? '}'` This would allow a graph term identifier such as `{=g}`, a graph definition including an identifier such as `{=g :s :p :o}`, or a graph term without an identifier `{ :s :p :o }`. This could work in Turtle, TriG and N-Quads (although N-Quads is limited to use the form without `triplesBlock?`).

Reply:
This syntactic feature seems possible to generalize into some `DEFINE SYMBOL VALUE` feature (to also handle e.g. repetition of strings; but it would be very similar to `PREFIX`...). Of course this is [dangerous territory](https://en.wikipedia.org/wiki/Billion_laughs_attack). Do we have a sense of which use cases call for it? (It might be indicative of an  issue with our design or in the data.) We should create a distinct issue/discussion for that.

Reply:
> some `DEFINE SYMBOL VALUE`

The [`USING %{label} FOR `](https://lists.w3.org/Archives/Public/public-rdf-star-wg/2023Oct/0017.html) is a "macro" (c.f. `#define`).

Other use cases are "large literals" - ones relating to content `rdf:HTML` , and `rdf:langString` and also `rdf:XMLIteral` and  `rdf:JSON`.RDF is already in "inverse dangerous territory" where content is forced to be written out repeatedly.

Would it be possible to take the use aside from graph terms to a separate issue?

Reply:
> Some syntax we might consider is `'{' ( '=' PN_CHARS+ )? triplesBlock? '}'`

Could you write out a complete example with definition and use?
(`{=g}` is a both definition and graph term identifier.)

I seem to remember that N3 did at one time have `:-` (or maybe it was `:=`) for a similar purpose on formula. That got taken out of cwm a very long time ago (well before W3C submission).

General point: I think we have to be careful which characters to use because viewed long term these characters are valuable and finite!

(I used words for clarity - I'm not arguing strongly that words are necessary.)

Reply:
(Aside: at one time we've used SPARQL construct with `VALUES` and an empty graph input as a way to do an "RDF macro" for repetitive chunks of data ([example](https://raw.githubusercontent.com/libris/definitions/develop/source/marc/construct-enums.rq)). A bit like CSVW but for more complex structures. i don't at all think Turtle should adopt something remotely as complex, but  `VALUE ?label "large term"` came to mind...)

Reply:
> > Some syntax we might consider is `'{' ( '=' PN_CHARS+ )? triplesBlock? '}'`
> 
> Could you write out a complete example with definition and use? (`{=g}` is a both definition and graph term identifier.)

Sure, consider something like the following:

```turtle
PREFIX : <http://example.com/#>
PREFIX rdfx: <http://example.com/rdfx#>

:a :referencesLabledGraph {=g1 :s1 :p :o1} .
:b :referencesLabledGraph {=g1 :s2 :p :o2} .
{=g1} a rdfx:GraphTerm .
:c :referencesUnlabeledGraph { :s3 :p :o3} .
```

The example defines a default graph containing four triples. A graph term (identified by _g1_) is the object of the first two triples and the subject of the third triple.

The first to statements define the same graph term, which collectively add two triples to it.

The third statement makes that graph term the subject of an `rdf:type`.

The fourth statement has another (unidentified) graph term as its object containing a single triple.

> I seem to remember that N3 did at one time have `:-` (or maybe it was `:=`) for a similar purpose on formula. That got taken out of cwm a very long time ago (well before W3C submission).

There was discussion of updating `blankNodePropertyList` to be able to start with `= subject`, such as `[ = :s :p :o] .` which would allow that `[ ... ]` form to be used more recursively. (May have been `:=` due to the problem the `=` is a verb, and so would create parsing issues).

> General point: I think we have to be careful which characters to use because viewed long term these characters are valuable and finite!

Sure, perhaps something other than `=` should be used here; it's not a problem for Turtle/TriG, but would be for N3.